internal/apiclient/pricing.go

```diff
@@ -169,26 +282,118 @@ func (c *PricingAPIClient) BatchRequests(resources []*schema.Resource, batchSize
 	return reqs
 }
 
+type pricingQuery struct {
+	hash  uint64
+	query GraphQLQuery
+
+	result gjson.Result
+}
+
+// PerformRequest sends a batch request to the Pricing API endpoint to fetch
+// pricing details for the provided queries. It optimizes the API call by
+// checking a local cache for previous results. If the results of a given query
+// are cached, they are used directly; otherwise, a request to the API is made.
 func (c *PricingAPIClient) PerformRequest(req BatchRequest) ([]PriceQueryResult, error) {
 	log.Debugf("Getting pricing details for %d cost components from %s", len(req.queries), c.endpoint)
+	res := make([]PriceQueryResult, len(req.keys))
+	for i, key := range req.keys {
+		res[i].PriceQueryKey = key
+	}
 
-	results, err := c.doQueries(req.queries)
+	queries := make([]pricingQuery, len(req.queries))
+	for i, query := range req.queries {
+		key, err := hashstructure.Hash(query, hashstructure.FormatV2, nil)
+		if err != nil {
+			logging.Logger.WithError(err).Debugf("failed to hash query %s will use nil hash", query)
+		}
+
+		queries[i] = pricingQuery{
+			hash:  key,
+			query: query,
+		}
+	}
+
+	// first filter any queries that have been stored in the cache. We don't need to
+	// send requests for these as we already have the results in memory.
+	var serverQueries []pricingQuery
+	if c.cache == nil {
+		serverQueries = queries
+	} else {
+		var hit int
+		for i, query := range queries {
+			v, ok := c.cache.Get(query.hash)
+			if ok {
+				logging.Logger.Debugf("cache hit for query hash: %d", query.hash)
+				hit++
+				res[i] = PriceQueryResult{
+					PriceQueryKey: req.keys[i],
+					Result:        v.Result,
+					filled:        true,
+				}
+			} else {
+				serverQueries = append(serverQueries, query)
+			}
+		}
+
+		logging.Logger.Debugf("%d/%d queries were built from cache", hit, len(queries))
+	}
+
+	// now we deduplicate the queries, ensuring that a request for a price only happens once.
+	var deduplicatedServerQueries []pricingQuery
+	seenQueries := map[uint64]bool{}
+	for _, query := range serverQueries {
+		if seenQueries[query.hash] {
+			continue
+		}
+
+		deduplicatedServerQueries = append(deduplicatedServerQueries, query)
+		seenQueries[query.hash] = true
+	}
+
+	// send the deduplicated queries to the pricing API to fetch live prices.
+	rawQueries := make([]GraphQLQuery, len(deduplicatedServerQueries))
+	for i, query := range deduplicatedServerQueries {
+		rawQueries[i] = query.query
+	}
+	resultsFromServer, err := c.doQueries(rawQueries)
```

Comment:
Not sure if this is easy or not, so might not be worth it. To reduce the round trips it would make more sense to me to do cache lookups and deduping before we batch the queries. So we only batch the queries we actually need to make requests for.

Reply:
yup I agree, but the way that the queries are tied to the `result`/`cost component` makes this tricky. As we expect the results to match the exact index of the cost component. I think it's something we can optimise later, but I'm going to leave out for now.